tests/Rule/Base/DifferentRuleInHandlerTestTrait.php

```diff
@@ -0,0 +1,31 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Validator\Tests\Rule\Base;
+
+use Yiisoft\Validator\Exception\UnexpectedRuleException;
+use Yiisoft\Validator\Tests\Support\Rule\RuleWithCustomHandler;
+use Yiisoft\Validator\Tests\Support\ValidatorFactory;
+
+trait DifferentRuleInHandlerTestTrait
+{
+    public function testDifferentRuleInHandler(): void
+    {
+        [$ruleClassName, $ruleHandlerClassName] = $this->getDifferentRuleInHandlerItems();
+
+        $rule = new RuleWithCustomHandler($ruleHandlerClassName);
+        $validator = ValidatorFactory::make();
+
+        $this->expectException(UnexpectedRuleException::class);
+        $this->expectExceptionMessage(
+            'Expected "' . $ruleClassName . '", but ' . RuleWithCustomHandler::class . ' given.'
+        );
+        $validator->validate([], [$rule]);
+    }
+
+    /**
+     * @psalm-return array{0:string, 1:string}
+     */
+    abstract protected function getDifferentRuleInHandlerItems(): array;
```

Comment:
Confusing name in my opinion, especially "different" part. Maybe `getRuleAndHandlerClassNames`?

Reply:
> Confusing name in my opinion, especially "different" part. Maybe `getRuleAndHandlerClassNames`?

This special method for test different rule in handler. This should be taken into account in the name.

Reply:
But aren't the rule and the handler always match tested ones?

Reply:
For example for `CountTest`:

```php
protected function getDifferentRuleInHandlerItems(): array
{
    return [Count::class, CountHandler::class];
}
```

So how they are "different"?

Reply:
Need check case, when to handler passed incorrect rule. See test `testDifferentRuleInHandler()`.